File: paymentsheet/src/test/java/com/stripe/android/paymentsheet/verticalmode/ShampooRule.java
package com.stripe.android.paymentsheet.verticalmode;

import org.junit.rules.TestRule;
import org.junit.runner.Description;
import org.junit.runners.model.Statement;

/** Got flaky tests? Shampoo them away. */
public final class ShampooRule implements TestRule {
    private final int iterations;

    public ShampooRule(int iterations) {
        if (iterations < 1) throw new IllegalArgumentException("iterations < 1: " + iterations);
        this.iterations = iterations;
    }

    @Override public Statement apply(final Statement base, Description description) {
        return new Statement() {
            @Override public void evaluate() throws Throwable {
                for (int i = 0; i < iterations; i++) {
                    base.evaluate();
                }
            }
        };
    }
}

File: payments-core-testing/src/main/java/com/stripe/android/testing/ShampooRule.kt
package com.stripe.android.testing

import org.junit.rules.TestRule
import org.junit.runner.Description
import org.junit.runners.model.Statement

/** Got flaky tests? Shampoo them away.  */
class ShampooRule(private val iterations: Int) : TestRule {
    init {
        require(iterations >= 1) { "iterations < 1: $iterations" }
    }

    override fun apply(base: Statement, description: Description?): Statement {
        return object : Statement() {
            @Throws(Throwable::class)
            override fun evaluate() {
                repeat(iterations) { iteration ->
                    try {
                        base.evaluate()
                    } catch (@Suppress("TooGenericExceptionCaught") e: Throwable) {
                        println("Failed on iteration: $iteration")
                        throw e
                    }
                }
            }
        }
    }
}
